Merging the plug-in of metadata database

mir-am opened the pull request:
I have created a new branch for the plug-in of the metadata database so that other members and Mihhail(@MihhailSokolov ) can contribute and commit to this PR. Please continue the review and discussion of the plug-in here.

For now, I do not see the insertion of metadata inside the `consume` method of the plug-in.
What is the current issue with the plug-in? is it JOOQ and database connection for CIs?

MihhailSokolov commented:
> For now, I do not see the insertion of metadata inside the `consume` method of the plug-in.
> What is the current issue with the plug-in? is it JOOQ and database connection for CIs?

Yes, I haven't yet implemented the consumption of data. The part that interacts with the database is completely done and tested.
Current issue is the problem of code generation in CI. I will rewrite CI to set up the correct environment with the database and appropriate schema which should fix the issue.

mir-am commented:
@MihhailSokolov 
Thanks for the work.
What is the current status of the plug-in?
Can we run it locally to insert records into the metadata DB?

MihhailSokolov commented:
> @MihhailSokolov
> Thanks for the work.
> What is the current status of the plug-in?
> Can we run it locally to insert records into the metadata DB?

@mir-am Yes, the plugin is ready and can be tested both locally and on the server.

gousiosg commented on analyzer/metadata-plugin/src/main/java/eu/fasten/analyzer/metadataplugin/MetadataDatabasePlugin.java:
```diff
@@ -0,0 +1,244 @@
+        @Override
+        public String description() {
+            return "Metadata plugin. "
+                    + "Consumes kafka topic and populates metadata database with consumed data.";
```
"Consumes _ExtendedRevisionCallgraph_-formatted JSON objects from a Kafka topic..._

MihhailSokolov replied:
Changed it

MihhailSokolov commented on analyzer/metadata-plugin/src/main/java/eu/fasten/analyzer/metadataplugin/Main.java:
```diff
@@ -0,0 +1,69 @@
+        }
+        try {
+            var metadataPlugin = new MetadataDatabasePlugin.MetadataPlugin();
+            var topic = "opal_callgraphs";
```
Added database connection parameters (URL, username, password) and topic name as command line arguments.

MihhailSokolov commented on analyzer/metadata-plugin/src/main/java/eu/fasten/analyzer/metadataplugin/MetadataDatabasePlugin.java:
```diff
@@ -0,0 +1,244 @@
+    @Extension
+    public static class MetadataPlugin implements KafkaConsumer<String> {
+
+        private final String topic = "opal_callgraphs";
```
I added topic parameter in the constructor of the MetadataPlugin.
Agree, having such method (like `getTopic()`) would be nice.

MihhailSokolov commented on analyzer/metadata-plugin/src/main/java/eu/fasten/analyzer/metadataplugin/MetadataDatabasePlugin.java:
```diff
@@ -0,0 +1,244 @@
+                        }
+                    });
+                } catch (JSONException e) {
+                    logger.error("Error parsing JSON callgraph", e);
```
Added coordinate name to the error message

MihhailSokolov commented on analyzer/metadata-plugin/src/main/java/eu/fasten/analyzer/metadataplugin/MetadataDatabasePlugin.java:
```diff
@@ -0,0 +1,244 @@
+                        if (getPluginError().isEmpty()) {
+                            processedRecord = true;
+                            restartTransaction = false;
+                            logger.info("Saved the callgraph metadata to the database");
```
Added coordinate name and the package ID from the database to the message

MihhailSokolov commented on analyzer/metadata-plugin/src/main/java/eu/fasten/analyzer/metadataplugin/MetadataDatabasePlugin.java:
```diff
@@ -0,0 +1,244 @@
+            int transactionRestartCount = 0;
+            do {
+                try {
+                    var callgraph = new ExtendedRevisionCallGraph(consumedJson);
```
I separated error handling for transactions and creating ERCG. However, I don't really see a point in having `continue` when catching transaction exception since it would be a last statement in the loop anyway. Maybe I just misunderstood you. Could you elaborate that part please @gousiosg ?

gousiosg replied:
You should continue to the next input object after you deal with the `JSONException`.

MihhailSokolov replied:
I extracted that part of the code (where I convert JSONObject into ERCG) out of the loop. So when it catches JSONException, it deals with it and returns from the method to consume the next record. Is this fine?

gousiosg replied:
Yes!

gousiosg commented:
Ready to merge

mir-am commented:
@MihhailSokolov 
Thanks for addressing the comments.
The PR is almost ready to merge. However, there are still two or three issues:

1. Main method of the plug-in throws the following exception when it's run with `mvn.dataset` call graphs but it seems it saves the call graphs' metadata into the DB (according to the logs.).
```
org.jooq.exception.TooManyRowsException: Cursor returned more than one result
	at org.jooq_3.12.3.POSTGRES.debug(Unknown Source)
	at org.jooq.impl.Tools.fetchOne(Tools.java:2087)
	at org.jooq.impl.AbstractResultQuery.fetchOne(AbstractResultQuery.java:653)
	at org.jooq.impl.SelectImpl.fetchOne(SelectImpl.java:2887)
	at eu.fasten.analyzer.metadataplugin.db.MetadataDao.getPackageIdByNameAndForge(MetadataDao.java:344)
	at eu.fasten.analyzer.metadataplugin.MetadataDatabasePlugin$MetadataPlugin.saveToDatabase(MetadataDatabasePlugin.java:170)
	at eu.fasten.analyzer.metadataplugin.MetadataDatabasePlugin$MetadataPlugin.lambda$consume$0(MetadataDatabasePlugin.java:116)
	at org.jooq.impl.DefaultDSLContext$3.run(DefaultDSLContext.java:608)
	at org.jooq.impl.DefaultDSLContext$3.run(DefaultDSLContext.java:605)
	at org.jooq.impl.DefaultDSLContext.lambda$transactionResult0$0(DefaultDSLContext.java:536)
	at org.jooq.impl.Tools$12$1.block(Tools.java:4843)
	at java.base/java.util.concurrent.ForkJoinPool.managedBlock(ForkJoinPool.java:3128)
	at org.jooq.impl.Tools$12.get(Tools.java:4840)
	at org.jooq.impl.DefaultDSLContext.transactionResult0(DefaultDSLContext.java:588)
	at org.jooq.impl.DefaultDSLContext.transactionResult(DefaultDSLContext.java:505)
	at org.jooq.impl.DefaultDSLContext.transaction(DefaultDSLContext.java:605)
	at eu.fasten.analyzer.metadataplugin.MetadataDatabasePlugin$MetadataPlugin.consume(MetadataDatabasePlugin.java:112)
	at eu.fasten.server.kafka.FastenKafkaConsumer.run(FastenKafkaConsumer.java:300)
```
2- The unit tests of the plug-in throws the following exception:
```
 ERROR eu.fasten.analyzer.metadataplugin.MetadataDatabasePlugin$MetadataPlugin - Error parsing JSON callgraph for 
org.json.JSONException: JSONObject["forge"] not found.
	at org.json.JSONObject.get(JSONObject.java:566)
	at org.json.JSONObject.getString(JSONObject.java:851)
	at eu.fasten.core.data.ExtendedRevisionCallGraph.<init>(ExtendedRevisionCallGraph.java:95)
	at eu.fasten.analyzer.metadataplugin.MetadataDatabasePlugin$MetadataPlugin.consume(MetadataDatabasePlugin.java:101)
	at eu.fasten.analyzer.metadataplugin.MetadataDatabasePluginTest.consumeJsonErrorTest(MetadataDatabasePluginTest.java:50)
	at java.base/jdk.internal.reflect.NativeMethodAccessorImpl.invoke0(Native Method)
	at java.base/jdk.internal.reflect.NativeMethodAccessorImpl.invoke(NativeMethodAccessorImpl.java:62)
	at java.base/jdk.internal.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43)
	at java.base/java.lang.reflect.Method.invoke(Method.java:566)
	at org.junit.platform.commons.util.ReflectionUtils.invokeMethod(ReflectionUtils.java:675)
	at org.junit.jupiter.engine.execution.ExecutableInvoker.invoke(ExecutableInvoker.java:115)
	at org.junit.jupiter.engine.descriptor.TestMethodTestDescriptor.lambda$invokeTestMethod$6(TestMethodTestDescriptor.java:170)
	at org.junit.jupiter.engine.execution.ThrowableCollector.execute(ThrowableCollector.java:40)
	at org.junit.jupiter.engine.descriptor.TestMethodTestDescriptor.invokeTestMethod(TestMethodTestDescriptor.java:166)
	at org.junit.jupiter.engine.descriptor.TestMethodTestDescriptor.execute(TestMethodTestDescriptor.java:113)
	at org.junit.jupiter.engine.descriptor.TestMethodTestDescriptor.execute(TestMethodTestDescriptor.java:58)
```
3- The FASTEN server finds the plug-in but it cannot load the plug-in as a Kafka consumer.

I will help to solve the third issue.

MihhailSokolov commented:
> The main method of the plug-in throws the following exception when it's run with mvn.dataset call graphs but it saves the call graphs' metadata into the DB (checked it by using SQL commands inside psql environment.).
I am going to look into that. You generated callgraphs from `mvn.dataset` and then ran `Main` class to consume from the `opal_callgraphs` topic?

> 2- The unit tests of the plug-in throws the following exception:
This is expected. I test there if it is going to throw an exception in case of incorrectly formatted JSON.

> The FASTEN server finds the plug-in but it cannot load the plug-in as a Kafka consumer. I will help to solve the third issue.
Thanks! I would appreciate your help on that.

MihhailSokolov commented:
@mir-am I fixed problem 1. Thanks for pointing it out!

mir-am commented:
@MihhailSokolov 
Thanks for fixing the issue.
I am working to solve the third issue, which is quite prominent. Otherwise, we cannot deploy the plug-in in the current pipeline we have.
